### ISSUE TITLE
BUG: pd.eval throwing - 'PandasExprVisitor' object has no attribute 'visit_JoinedStr'

### ISSUE THREAD

**Issue by dss010101** [pandas-dev/pandas#52310]
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd, numpy as np
import timeit
class Data:
    def __init__(self, data, tag = 'xxx') -> None:
        self.data = data
        self.tag1 =  'xxx'
        
df = pd.DataFrame(np.random.randint(0,10,100).astype(str).reshape(50,2),columns=list('AB'))
d = Data(data=df)

eval("data.A + f':{tag1}'", d.__dict__)

pd.eval("data.A + f':{tag1}'", resolvers=(d.__dict__,))
```

### Issue Description

In the above, eval works fine.  However pd.eval  throws the following exception

AttributeError: 'PandasExprVisitor' object has no attribute 'visit_JoinedStr'

### Expected Behavior

If eval is able to evaluate this correctly, then pandas.eval should i feel.

For context, the reason for code like this, is that f strings dont seem to work with pandas series.  so i cant do something like:

### Installed Versions

<details>

Replace this line with the output of pd.show_versions()

</details>

**Comment by topper-123** [comment]
We haven't implemented f-string in pd.eval. Couldn't you  add the `xxx` variable directly into the string?

**Comment by dss010101** [comment]
> We haven't implemented f-string in pd.eval. Couldn't you add the `xxx` variable directly into the string?

yes ofcourse. but this example is greatly simplified to demonstrate the issue.  the actual use case a more generic piece of code that deals with expression evaluation for inputs may be python popo's or pandas dataframe

**Comment by topper-123** [comment]
The string inside pd.eval isn't normal python code like the plan eval, because it supports e.g. @ and backslashes. Can you give an example wouldn't work/be inconvenient?

**Comment by dss010101** [comment]
IIt's been a while since i opened the ticket so i can not recall exactly the use cases that was failing...
but we have configuration file that defines calculated fields by using an expression.  the expression may refer to datasets + columns (i.e. df.A above), or other variables, either single variables/constants or variables that can be object properties or even function calls.  
The scope of known dataset is a python class representing the data model or domain - so you can build an expression by using any properties or methods on that data model.  Again those properties may be single values, constants or timeseries (dataframes).

an example would be something like:
```
{
    "calc_field1",  
     expression: "{port.spread} * {port.mkt_val} + ( {rolling_avg(200, rate)} * 100)"

    "calculed_fiedl2",  
     expression: "{port.name}_batch_{batch__id}"
}

i assume everything in curly braces can be evaluated against the data model instance, and everything outside is simply a constant.  so after little parsing/massaging (if need be),  i generate a string to be evaluated


```

**Comment by topper-123** [comment]
I think your example is geared towards `str.format` and passing that to python's `eval` (which certainly can be a fine strategy, nothing wrong with that).

A setup using pd.eval would end up like something like `pd.eval("port.spread * port.mkt_val + (rolling_avg(200, rate) / 100)", local_dict=dict(port=df, rolling_avg=rolling_avg, rate=rate))` or similar.

I think passing variables to local_dict is the way to go for `pd.eval` and should give the needed end result.

**Comment by dss010101** [comment]
i get what you're saying. in our case str.format would not be a **straight forward/simple** solution since we want to assign back to a dataframe and depending on the calculation, the calculated field could be numeric or string. also in our case, we pass the class model  as the dict, and tell the expression writer they can use any properties or methods on that model in their expression (there may be different models), so we wanted to avoid specifying individual properties/method as a local_dict in the 'template/expression engine'

I think at the end of the day, it boils down to simplicity...writing less code, cleanly and more readable.. for example below all 3 evals solve the problem..but i much prefer the succintness of the 3rd and the flexibility it offers

```
import pandas as pd, numpy as np
import timeit
class DataModel:
    def __init__(self, data, tag = 'xxx') -> None:
        self.data = data
        self.tag1 =  'tag_1'
        self.tag2 =  'tag_2'
        self.tag3 =  'tag_3'
        self.id_tmplt = '{tag1}_{tag2}_{tag3}_id' 
        
df = pd.DataFrame(np.random.randint(0,10,10).reshape(5,2),columns=list('AB'))
d = DataModel(data=df)

df['id'] = eval("data.A.astype(str) + tag1 + '_' + tag2 + '_' + tag3 + '_id'", d.__dict__)
df['id'] = eval("data.A.astype(str) + id_tmplt.format(tag1=tag1, tag2=tag2, tag3=tag3)", d.__dict__)
df['id'] = eval("data.A.astype(str) + f'{id_tmplt}'", d.__dict__)

df
```
